### PR TITLE
[Feature] support in predict for complex types

### PR DESCRIPTION
**Pull request by fzhedu**
1. array/map/struct support in (value..) and in (subquery), json just support in (value...)
2. fix map equal bugs
3. fix stuct type's getCompatibleTypes error
4. support not equal for array/map/struct
5. fix cast struct bug of lacking nullable

## What type of PR is this:
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

File: be/src/exprs/in_const_predicate.hpp [hunk 1]
```diff
@@ -395,6 +399,120 @@ class VectorizedInConstPredicate final : public Predicate {
...
+        Columns input_data(child_size);
+        std::vector<NullColumnPtr> input_null(child_size);
+        std::vector<bool> is_const(child_size);
+        Columns columns_ref(child_size);
+        ColumnPtr value;
+        bool all_const = true;
+        for (int i = 0; i < child_size; ++i) {
+            value = _const_input[i];
+            if (value == nullptr) {
+                ASSIGN_OR_RETURN(value, _children[i]->evaluate_checked(context, ptr));
```
Comment (Seaven) [inline be/src/exprs/in_const_predicate.hpp:438]: the children of `in` must be constant, I think you can add DCHECK on `expr.is_const`,  and don't need complex handle

File: be/src/exprs/in_const_predicate.hpp [hunk 2]
```diff
@@ -395,6 +399,120 @@ class VectorizedInConstPredicate final : public Predicate {
...
+    ~VectorizedInConstPredicateGeneric() override = default;
+
+    Expr* clone(ObjectPool* pool) const override { return pool->add(new VectorizedInConstPredicateGeneric(*this)); }
+
+    Status open(RuntimeState* state, ExprContext* context, FunctionContext::FunctionStateScope scope) override {
+        RETURN_IF_ERROR(Expr::open(state, context, scope));
+        _const_input.resize(_children.size());
+        for (auto i = 0; i < _children.size(); ++i) {
+            if (dynamic_cast<VectorizedLiteral*>(_children[i])) {
+                ASSIGN_OR_RETURN(_const_input[i], _children[i]->evaluate_checked(context, nullptr));
```
Comment (Seaven) [inline be/src/exprs/in_const_predicate.hpp:419]: the children of `in` must be constant, I think you can add DCHECK on `expr.is_const`,  and don't need complex handle

Comment (fzhedu) [reply be/src/exprs/in_const_predicate.hpp:419]: seems `Expr::is_constant()` is not overrided by any expression.

Comment (fzhedu) [reply be/src/exprs/in_const_predicate.hpp:419]: as currently rewritting non-const parameters of in predict to OR  expressions results in slower performance, we could optimize it by allow in predict to support non-const parameteres, so it is necessary to keep this generial process.

Comment (Seaven) [reply be/src/exprs/in_const_predicate.hpp:419]: > seems `Expr::is_constant()` is not overrided by some expression, and this interface is not really used.

1. only a few functions doesn't support `is_constant`, such as `random`, don't worry.
2. as you say,  I think rewrite contant expression on `IN` predicate is unnecessary, but it's only need to modify FE. you can think about it, execut a constant expression once more is faster than execute it as a variable column in your template

Comment (Seaven) [reply be/src/exprs/in_const_predicate.hpp:419]: BTW, support variable column on `IN` is not necessarily faster than using OR

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 9 / 13 (69.23%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|3|6|50.00%|[800, 801, 807]|
|:large_blue_circle:|com/starrocks/analysis/InPredicate.java|2|3|66.67%|[135]|
|:large_blue_circle:|com/starrocks/sql/common/TypeManager.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|3|3|100.00%|[]|

**Review by Seaven — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 9 / 13 (69.23%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|3|6|50.00%|[800, 801, 807]|
|:large_blue_circle:|com/starrocks/analysis/InPredicate.java|2|3|66.67%|[135]|
|:large_blue_circle:|com/starrocks/sql/common/TypeManager.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|3|3|100.00%|[]|

**Review by murphyatwork — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 9 / 13 (69.23%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|3|6|50.00%|[800, 801, 807]|
|:large_blue_circle:|com/starrocks/analysis/InPredicate.java|2|3|66.67%|[135]|
|:large_blue_circle:|com/starrocks/sql/common/TypeManager.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|3|3|100.00%|[]|

File: be/src/column/map_column.cpp [hunk 3]
```diff
@@ -422,34 +423,75 @@ int MapColumn::equals(size_t left, const Column& rhs, size_t right, bool safe_eq
...
-            has_null |= (res == EQUALS_NULL);
-            // So two keys is the same
-            res = _values->equals(i, *(rhs_map._values.get()), j, safe_eq);
-            if (res == EQUALS_FALSE) {
-                return EQUALS_FALSE;
+            // So two keys are the same or right key is null
+            int val_res = _values->equals(i, *(rhs_map._values.get()), j, safe_eq);
+
+            // case 1: key_res == EQUALS_TRUE
+            if (key_res == EQUALS_TRUE) {
```
Comment (chaoyli) [inline be/src/column/map_column.cpp:463]: The `if (key_res == EQUALS_TRUE) {` is non-sense.

Comment (fzhedu) [reply be/src/column/map_column.cpp:463]: not true, it has to diff the case where `key_res == EQUAL_NULL`

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 9 / 13 (69.23%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|3|6|50.00%|[800, 801, 807]|
|:large_blue_circle:|com/starrocks/analysis/InPredicate.java|2|3|66.67%|[135]|
|:large_blue_circle:|com/starrocks/sql/common/TypeManager.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 9 / 13 (69.23%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|3|6|50.00%|[800, 801, 807]|
|:large_blue_circle:|com/starrocks/analysis/InPredicate.java|2|3|66.67%|[135]|
|:large_blue_circle:|com/starrocks/sql/common/TypeManager.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|3|3|100.00%|[]|

**Review by murphyatwork — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 9 / 13 (69.23%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|3|6|50.00%|[800, 801, 807]|
|:large_blue_circle:|com/starrocks/analysis/InPredicate.java|2|3|66.67%|[135]|
|:large_blue_circle:|com/starrocks/sql/common/TypeManager.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|3|3|100.00%|[]|

**Comment by fzhedu** [comment]
@mergify backport branch-3.1

**Pull request merged**